storage/badger/receipts.go

```diff
@@ -143,7 +143,10 @@ func (r *ExecutionReceipts) Index(blockID, receiptID flow.Identifier) error {
 	})
 }
 
-func (r *ExecutionReceipts) IndexByBlockIDAndExecutionID(blockID, executorID, receiptID flow.Identifier) error {
+func (r *ExecutionReceipts) IndexByExecutor(receipt *flow.ExecutionReceipt) error {
```

Comment:
Agree with your points.

For the first one, I think the current strategy to protect from BFT is to crash the consensus node, as soon as we notice there is conflicting results either from different nodes or same node. If we continue to follow this strategy, then we could apply a simple fix, which is to check if we have already stored a receipt from this executor, and if the received one is different from the stored one, crash the consensus node. This prevents from sealing a wrong result.

For the second point, we could add the check to make sure the indexed receipt exist. And remove the check on the result. I agree it's unnecessary.

Reply:
>If that operation succeeds, we don't check whether the receipt's full text was stored in the first place. The algorithm happily returns without error if the receipt itself is unknown (which would lead to downstream bugs when trying to retrieve the receipts)

When indexing a receipt, we actually checked if there is already an index, and if the index exists, it checks if the stored receipts also exist. [An error will return](https://github.com/onflow/flow-go/pull/387/files#diff-b20e9c1487d662b022af6e66fde72fa5d42d657195e1cc53cef20cfe90180d58R173) if the index exists, the the receipt was missing

Reply:
This is new functionality we are implementing. Lets please do it properly and not create too much tech debt. 
* the previous version of the `storage.ExecutionReceipts` worked well for what the execution node needed.
* In your PR, you are [adding API components](https://github.com/onflow/flow-go/blob/93c34ee74265d90163dfbe451997fc8491b17399/storage/receipts.go#L23-L36), which have intrinsically conflicting use cases to the existing one. 

I think it would be reasonable to split the functionality of the `storage.ExecutionReceipts` and make a version for the consensus nodes which handles conflicting receipts properly.  


#### Regarding: support multiple different receipts from the same execution node for the same block

I have the feeling that we are potentially talking about slightly different things here:
* I agree with your point that currently, we need to prevent two different execution results being sealed. Currently, this safeguard mechanism is implemented as a wrapper [`ExecForkSuppressor`](https://github.com/onflow/flow-go/blob/93c34ee74265d90163dfbe451997fc8491b17399/module/mempool/consensus/exec_fork_suppressor.go).
* My point is: our implementation should generally handle multiple receipts from the same EN as this is a requirement for the mature protocol. 
   - You are right that a different part of the system does not support conflicting receipts at the moment
   - But that doesn't mean that the storage layer doesn't need to support it either. We are working hard to update the parts of the implementation which can't support conflicting receipts. So lets not add other systems to that tech debt, but rather implement the new logic, so it supports it right away. 

#### Regarding: the second point
I should have worded this different: its _not_ about deduplication, its about the ability to indexing non-stored receipts. That should not be allowed in my opinion.

Reply:
To be addressed in downstream PR. Issue'ed up work: https://github.com/dapperlabs/flow-go/issues/5326